### DISCUSSION TITLE
Primitives as npm package or copy/pasted?

### DISCUSSION THREAD

**Discussion by mrzachnugent** [founded-labs/react-native-reusables#49] [Polls]
# How would you like to add primitives to your project?

## Option 1: NPM package

Like ui/shadcn, the primitives would be an npm package that you install in your project so that the code lives in your `node_modules`. This would be one less thing to think about. However, since this project is still in its early stages, fixing bugs would require a patch-package or submitting an issue/pull-request and waiting for the code to be fixed.

## Option 2: Copy/pasted

Instead of the code living in the `node_modules` folder, you would have a folder called `primitives`. Like your `ui` folder, this is code that you would rarely have to touch. It could be helpful if you want to customize the primitives. For example, you could use a different state manager like redux or zustand. This would also give you the freedom to only add the native primitives if you do not want to use the web primitives.

### Your experience with ui/shadcn VS react-native-reusables

If you'd like to leave a comment about your experience, here are some topics you could share about:
- How often you need to change the code in your `ui` folder?
- Is it something you set up once and rarely think about? 
- Would adding another folder for `primitives` hinder your developer experience?

**Comment by Linzo99** [comment]
Like you said, since this project is still in its early stages, i think it is more efficient to have it as a folder in a sense that it will help for faster updates but also allow customizations.

But for those who find it stable enough maybe it's easier to have it as a package, for example i can hardly find a situation where i would like to manually change "radix primitives".

So maybe it wouldn't hurt having both methods in the future.

**Comment by corysimmons** [comment]
For sure an npm package.

These are supposed to mimic Radix UI Primitives which are npm packages and I don't think anyone has _ever_ had to customize those.

I'd also put them in their own package (maybe even repo?) to keep clutter down and focus the responsibility of this project to mostly just the components.

This also means people can copy folders directly from the repo since the primitives won't be in the same repo, but rather actually from an npm package.

**Comment by mrzachnugent** [comment]
# New option

A [CLI](https://github.com/mrzachnugent/react-native-reusables/tree/main/packages/cli) to add components to your projects. It does require you to follow the [initial setup](https://rnr-docs.vercel.app/getting-started/initial-setup/) before use.

Let me know what you think about the CLI ( **very early stage**)

**Reply by corysimmons** [reply]
Will this automatically install primitives for the components we want to use?

**Reply by mrzachnugent** [reply]
@corysimmons Yes exactly.

It will add all of the required primitives and dependencies.

The CLI will also ask if it is for a universal (web,iOS, and Android) or a native-only (iOS and Android) project so that it can add the files and dependencies accordingly.

**Reply by corysimmons** [reply]
Nice! That's exactly what I've been waiting for.

**Comment by mrzachnugent** [comment]
Now available as npm packages https://rn-primitives.vercel.app/  🥳

_You can still copy/paste them if you want_
